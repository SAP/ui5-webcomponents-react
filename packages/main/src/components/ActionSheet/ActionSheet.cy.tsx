import { Button, ButtonPropTypes, Label } from '../../webComponents';
import { ActionSheet, ActionSheetPropTypes } from './index';

interface TestCompProptypes extends ActionSheetPropTypes {
  onBtnClick?: ButtonPropTypes['onClick'];
}

const TestComp = (props: TestCompProptypes) => {
  const { onBtnClick, open, children } = props;
  return (
    <>
      <button id="opener">Opener</button>
      <ActionSheet className="myCustomClass" open={open} opener="opener">
        {children || (
          <>
            <Button onClick={onBtnClick}>Accept</Button>
            <Button>Reject</Button>
            <Button>This is my super long text!</Button>
          </>
        )}
      </ActionSheet>
    </>
  );
};

describe('ActionSheet', () => {
  it('Click Action', () => {
    const onBtnClick = cy.spy().as('onBtnClick');
    cy.mount(<TestComp onBtnClick={onBtnClick} open />);

    cy.get('[ui5-responsive-popover]').should('be.visible');

    cy.findByText('Accept').click();
    cy.get('[ui5-responsive-popover]').should('not.be.visible');
    cy.get('@onBtnClick').should('have.been.calledOnce');
  });

  it('does not crash with other component', () => {
    cy.mount(
      <TestComp open>
        <Label>I should not crash</Label>
      </TestComp>
    );
    cy.findByText('I should not crash').should('be.visible');
  });
});
