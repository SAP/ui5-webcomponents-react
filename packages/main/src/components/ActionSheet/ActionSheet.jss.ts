import { ThemingParameters, CssSizeVariables } from '@ui5/webcomponents-react-base';

const styles = {
  /*
   *  is being applied to the encapsulating ul element
   */
  actionSheet: {
    '&::part(content)': {
      padding: '0.1875rem 0.375rem'
    },
    '& [ui5-button]': {
      margin: CssSizeVariables.sapWcrButtonTopBottomPadding,
      display: 'block'
    },
    '& [ui5-button]:not([data-cancel-btn])::part(button)': {
      justifyContent: 'start'
    },
    '&::part(header)': {
      background: 'transparent',
      boxShadow: 'none'
    }
  },
  actionSheetMobile: {
    '&::part(content)': {
      padding: '0.1875rem 0.375rem 0 0.375rem'
    }
  },
  contentMobile: {
    backgroundColor: ThemingParameters.sapGroup_ContentBackground,
    padding: '0.25rem 0.5rem',
    boxSizing: 'border-box',
    overflow: 'auto',
    maxHeight: '82vh',
    borderTopLeftRadius: ThemingParameters.sapElement_BorderCornerRadius,
    borderTopRightRadius: ThemingParameters.sapElement_BorderCornerRadius,
    boxShadow: ThemingParameters.sapContent_Shadow2,
    '& > *': {
      margin: '0.25rem 0'
    }
  }
};

export default styles;
