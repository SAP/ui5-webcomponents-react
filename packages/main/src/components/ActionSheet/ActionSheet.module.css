.actionSheet::part(content) {
  padding: 0.1875rem 0.375rem;
}

.actionSheet [ui5-button] {
  margin: var(--_ui5wcr-ButtonTopBottomPadding);
  display: block;
}

.actionSheet [ui5-button]:not([data-cancel-btn])::part(button) {
  justify-content: flex-start;
}

.actionSheet::part(header) {
  background: transparent;
  box-shadow: var(--_ui5wcr_ActionSheet_HeaderBoxShadow);
}

/* actionSheetMobile */
.actionSheetMobile::part(content) {
  padding: 0 !important;
}

/* contentMobile */
.contentMobile {
  background-color: var(--sapGroup_ContentBackground);
  padding: 0.25rem 0.5rem;
  box-sizing: border-box;
  overflow: auto;
  max-height: 82vh;
  border-top-left-radius: var(--_ui5wcr_ActionSheet_TopBorderRadius);
  border-top-right-radius: var(--_ui5wcr_ActionSheet_TopBorderRadius);
  box-shadow: var(--_ui5wcr_ActionSheet_BoxShadow);
}

.contentMobile > * {
  margin: 0.25rem 0;
}
