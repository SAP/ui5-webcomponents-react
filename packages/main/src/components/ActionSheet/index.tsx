import { isPhone } from '@ui5/webcomponents-base/dist/Device.js';
import { addCustomCSS } from '@ui5/webcomponents-base/dist/Theming.js';
import { useI18nBundle, useSyncRef } from '@ui5/webcomponents-react-base/dist/hooks';
import { ThemingParameters } from '@ui5/webcomponents-react-base/dist/ThemingParameters';
import { AVAILABLE_ACTIONS, CANCEL, X_OF_Y } from '@ui5/webcomponents-react/dist/assets/i18n/i18n-defaults';
import { Button } from '@ui5/webcomponents-react/dist/Button';
import { ButtonDesign } from '@ui5/webcomponents-react/dist/ButtonDesign';
import { ResponsivePopover } from '@ui5/webcomponents-react/dist/ResponsivePopover';
import clsx from 'clsx';
import React, {
  Children,
  cloneElement,
  forwardRef,
  ReactElement,
  RefObject,
  useCallback,
  useReducer,
  useRef
} from 'react';
import { createPortal } from 'react-dom';
import { createUseStyles } from 'react-jss';
import { ButtonPropTypes } from '../../webComponents/Button';
import { ResponsivePopoverDomRef, ResponsivePopoverPropTypes } from '../../webComponents/ResponsivePopover';
import styles from './ActionSheet.jss';

export interface ActionSheetPropTypes extends Omit<ResponsivePopoverPropTypes, 'children'> {
  /**
   * Defines the actions of the `ActionSheet`.
   *
   * __Note:__ Although this slot accepts all HTML Elements, it is strongly recommended that you only use `Buttons` in order to preserve the intended design.
   */
  children?: ReactElement<ButtonPropTypes> | ReactElement<ButtonPropTypes>[];
  /**
   * Displays a cancel button below the action buttons on mobile devices. No cancel button will be shown on desktop and tablet devices.
   */
  showCancelButton?: boolean;
  /**
   * Defines whether the `header` or `headerText` should always be displayed or only on mobile devices.
   */
  alwaysShowHeader?: boolean;
  /**
   * Defines internally used a11y properties.
   */
  a11yConfig?: {
    actionSheetMobileContent?: {
      role?: string;
      ariaLabel?: string;
    };
  };
  /**
   * Defines where modals are rendered into via `React.createPortal`.
   *
   * Defaults to: `document.body`
   */
  portalContainer?: Element;
}

const useStyles = createUseStyles(styles, { name: 'ActionSheet' });

if (isPhone()) {
  addCustomCSS(
    'ui5-responsive-popover',
    `
  :host([data-actionsheet]) [ui5-button] {
    display: none;
  }
  :host([data-actionsheet]) [ui5-dialog] {
    top: auto !important;
    bottom: 0;
    height: auto;
    border-radius: 0;
    background-color: transparent;
    box-shadow: none;
    box-sizing: border-box;
    min-height: unset;
  }
  :host([data-actionsheet]) [ui5-title] {
    color: ${ThemingParameters.sapContent_ContrastTextColor} !important;
    text-shadow: none;
    text-align: start !important;
    padding-left:1rem !important;
  }
  `
  );
}

/**
 * The `ActionSheet` holds a list of buttons from which the user can select to complete an action. <br />
 * The children of the action sheet should be `Button` components. Elements in the `ActionSheet` are left-aligned. Actions should be arranged in order of importance, from top to bottom.
 */
const ActionSheet = forwardRef((props: ActionSheetPropTypes, ref: RefObject<ResponsivePopoverDomRef>) => {
  const {
    a11yConfig,
    allowTargetOverlap,
    alwaysShowHeader,
    children,
    className,
    footer,
    header,
    headerText,
    hideArrow,
    horizontalAlign,
    initialFocus,
    modal,
    placementType,
    portalContainer,
    showCancelButton,
    slot,
    style,
    verticalAlign,
    onAfterClose,
    onAfterOpen,
    onBeforeClose,
    onBeforeOpen,
    ...rest
  } = props;
  const i18nBundle = useI18nBundle('@ui5/webcomponents-react');
  const classes = useStyles();
  const [componentRef, popoverRef] = useSyncRef(ref);
  const actionBtnsRef = useRef(null);
  const [focusedItem, setFocusedItem] = useReducer((_, action) => {
    return parseInt(action.target.dataset.actionBtnIndex);
  }, 0);
  const childrenToRender = Children.toArray(children).filter(Boolean);
  const childrenArrayLength = childrenToRender.length;
  const childrenLength = isPhone() && showCancelButton ? childrenArrayLength + 1 : childrenArrayLength;

  const handleCancelBtnClick = () => {
    popoverRef.current.close();
  };

  const onActionButtonClicked = (handler) => (e) => {
    popoverRef.current.close();
    if (typeof handler === 'function') {
      handler(e);
    }
  };

  const handleFocus = (handler) => (e) => {
    if (typeof handler === 'function') {
      handler(e);
    }
    setFocusedItem(e);
  };

  const handleKeyDown = useCallback(
    (e) => {
      const currentIndex = parseInt(e.target.dataset.actionBtnIndex);
      if (e.key === 'ArrowDown' && currentIndex + 1 < childrenLength) {
        actionBtnsRef.current.querySelector(`[data-action-btn-index="${currentIndex + 1}"]`).focus();
      }
      if (e.key === 'ArrowUp' && currentIndex > 0) {
        actionBtnsRef.current.querySelector(`[data-action-btn-index="${currentIndex - 1}"]`).focus();
      }
    },
    [childrenLength, actionBtnsRef.current]
  );

  const renderActionSheetButton = (element, index: number, childrenArray) => {
    return cloneElement(element, {
      role: 'button',
      'aria-label': `${i18nBundle.getText(X_OF_Y, index + 1, childrenArray.length)} ${element.props?.children}`,
      key: index,
      tabIndex: focusedItem === index ? 0 : -1,
      ...element.props,
      design: ButtonDesign.Transparent,
      onClick: onActionButtonClicked(element.props?.onClick),
      onFocus: handleFocus(element.props?.onFocus),
      'data-action-btn-index': index
    });
  };

  const handleAfterOpen = useCallback(
    (e) => {
      if (isPhone()) {
        actionBtnsRef.current.querySelector(`[data-action-btn-index="${focusedItem}"]`).focus();
      }
      if (typeof onAfterOpen === 'function') {
        onAfterOpen(e);
      }
    },
    [onAfterOpen, actionBtnsRef.current, focusedItem]
  );
  const displayHeader = alwaysShowHeader || isPhone();
  return createPortal(
    <ResponsivePopover
      aria-modal
      role="dialog"
      style={style}
      slot={slot}
      allowTargetOverlap={allowTargetOverlap}
      headerText={displayHeader ? headerText : undefined}
      horizontalAlign={horizontalAlign}
      initialFocus={initialFocus}
      modal={modal}
      hideArrow={hideArrow}
      placementType={placementType}
      verticalAlign={verticalAlign}
      footer={footer}
      header={displayHeader ? header : undefined}
      onAfterClose={onAfterClose}
      onBeforeClose={onBeforeClose}
      onBeforeOpen={onBeforeOpen}
      {...rest}
      onAfterOpen={handleAfterOpen}
      ref={componentRef}
      className={clsx(classes.actionSheet, className)}
      data-actionsheet
    >
      <div
        className={isPhone() ? classes.contentMobile : undefined}
        data-component-name="ActionSheetMobileContent"
        role={a11yConfig?.actionSheetMobileContent?.role ?? 'presentation'}
        aria-label={a11yConfig?.actionSheetMobileContent?.ariaLabel ?? i18nBundle.getText(AVAILABLE_ACTIONS)}
        onKeyDown={handleKeyDown}
        ref={actionBtnsRef}
      >
        {childrenToRender.map(renderActionSheetButton)}
        {isPhone() && showCancelButton && (
          <Button
            design={ButtonDesign.Negative}
            onClick={handleCancelBtnClick}
            tabIndex={focusedItem === childrenLength - 1 ? 0 : -1}
            data-action-btn-index={childrenLength - 1}
            data-cancel-btn
            onFocus={setFocusedItem}
          >
            {i18nBundle.getText(CANCEL)}
          </Button>
        )}
      </div>
    </ResponsivePopover>,
    portalContainer
  );
});

ActionSheet.defaultProps = {
  showCancelButton: true,
  alwaysShowHeader: true,
  portalContainer: document.body
};

ActionSheet.displayName = 'ActionSheet';

export { ActionSheet };
