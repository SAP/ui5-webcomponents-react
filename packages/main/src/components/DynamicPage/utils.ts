export const DynamicPageCssVariables = {
  headerDisplay: '--_ui5wcr_DynamicPage_header_display',
  titleFontSize: '--_ui5wcr_DynamicPage_title_fontsize'
};
