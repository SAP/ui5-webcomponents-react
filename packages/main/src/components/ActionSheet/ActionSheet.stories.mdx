import { ArgsTable, Canvas, Meta, Story } from '@storybook/addon-docs';
import { ActionSheet } from '@ui5/webcomponents-react/dist/ActionSheet';
import { Button } from '@ui5/webcomponents-react/dist/Button';
import { PopoverPlacementType } from '@ui5/webcomponents-react/dist/PopoverPlacementType';
import { PopoverHorizontalAlign } from '@ui5/webcomponents-react/dist/PopoverHorizontalAlign';
import { PopoverVerticalAlign } from '@ui5/webcomponents-react/dist/PopoverVerticalAlign';
import { useCallback, useRef } from 'react';
import { DocsHeader } from '@shared/stories/DocsHeader';

<Meta
  title="Modals & Popovers / ActionSheet"
  component={ActionSheet}
  argTypes={{
    children: {
      type: null
    },
    footer: {
      type: null
    },
    header: {
      type: null
    }
  }}
  args={{
    horizontalAlign: PopoverHorizontalAlign.Center,
    placementType: PopoverPlacementType.Right,
    verticalAlign: PopoverVerticalAlign.Center
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      const actionSheetRef = useRef();
      const onButtonClick = useCallback(
        (e) => {
          actionSheetRef.current.showAt(e.target);
        },
        [actionSheetRef]
      );
      return (
        <>
          <Button onClick={onButtonClick}>Open ActionSheet</Button>
          <ActionSheet {...args} ref={actionSheetRef}>
            <Button icon="add">Accept</Button>
            <Button>Reject</Button>
            <Button>This is my super long text!</Button>
          </ActionSheet>
        </>
      );
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />

## Methods

This component exposes public methods. You can invoke them directly on the instance of the component, e.g. by using React Refs.

| Name           | Parameters                                                                                                                                                    | Description                                                                                                      |
| -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------- |
| **close**      | &mdash;                                                                                                                                                       | Closes the popover/dialog.                                                                                       |
| **open**       | <dl><dt className="methodText">**opener**</dt><dd className="methodText">the element that the popover is opened by</dd><dd><code>HTMLElement</code></dd></dl> | Opens popover on desktop and dialog on mobile.                                                                   |
| **applyFocus** | &mdash;                                                                                                                                                       | Focuses the element denoted by <code>initialFocus</code>, if provided, or the first focusable element otherwise. |
| **isOpen**     | &mdash;                                                                                                                                                       |

## Opening ActionSheets

`ActionSheets` can only be opened by attaching a `ref` to the component and then call the corresponding **`open`** method. The method receives the target element - _on which the `ActionSheet` is to be opened_ - as parameter.

```JSX
const ActionSheetComponent = () => {
  const actionSheetRef = useRef();
  const onButtonClick = (e) => {
    actionSheetRef.current.showAt(e.target);
  };
  return (
    <>
      <Button onClick={onButtonClick}>Open ActionSheet</Button>
      <ActionSheet ref={actionSheetRef}>
        <Button icon="add">Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    </>
  );
};
```

## Using ActionSheets inside other components

`ActionSheets` are often used within other components, when opened this could sometimes have unwanted side effects.
In this case, we recommend using [createPortal](https://reactjs.org/docs/portals.html) to mount the `ActionSheet` outside of the DOM hierarchy of the parent component.

```JSX
const ActionSheetComponent = () => {
  const actionSheetRef = useRef();
  const onButtonClick = (e) => {
    actionSheetRef.current.showAt(e.target);
  };
  return (
    <>
      <Button onClick={onButtonClick}>Open ActionSheet</Button>
      {createPortal(
        <ActionSheet ref={actionSheetRef}>
          <Button icon="add">Accept</Button>
          <Button>Reject</Button>
          <Button>This is my super long text!</Button>
        </ActionSheet>,
        document.body
      )}
    </>
  );
};
```
