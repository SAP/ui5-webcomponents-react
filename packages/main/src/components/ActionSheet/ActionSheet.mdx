import { ControlsWithNote, DocsHeader, DomRefTable, Footer } from '@sb/components';
import { Canvas, Meta } from '@storybook/blocks';
import { MessageStrip } from '../../webComponents/index';
import * as ComponentStories from './ActionSheet.stories';
import ResponsivePopoverDomRef from '../../webComponents/ResponsivePopover/ResponsivePopoverDomRef.json';

<Meta of={ComponentStories} />

<DocsHeader />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<DomRefTable rows={ResponsivePopoverDomRef} /> |

## Opening ActionSheets

<br />

#### since 0.22.0

We recommend opening and closing the `ActionSheet` component in a declarative way using the `open` and `opener` prop.
You can still use the imperative way which is outlined below.

```jsx
const MyComponentWithActionSheet = () => {
  const [actionSheetIsOpen, setActionSheetIsOpen] = useState(false);
  return (
    <>
      <Button
        id={'openActionSheetBtn'}
        onClick={() => {
          setActionSheetIsOpen(true);
        }}
      >
        Open Action Sheet
      </Button>
      <ActionSheet
        opener={'openActionSheetBtn'}
        open={actionSheetIsOpen}
        onAfterClose={() => {
          setActionSheetIsOpen(false);
        }}
      />
    </>
  );
};
```

**Opening an `ActionSheet` by reference and not by `id`**

This web component exposes a way to pass a reference of an element instead of an `id` to the `opener` prop. Since this is not supported when passing the reference in a declarative way to a React `prop`, you have to attach the ref directly on the web component.
You can do that by e.g. leveraging a React Ref, and then set the corresponding property there.

<MessageStrip
  hideCloseButton
  children={`Changing properties outside of the React rendering cycle can lead to unwanted behavior, please be extra cautious when
  using this approach!`}
/>

```jsx
const ActionSheetComponent = () => {
  const popoverRef = useRef(null);
  const [open, setOpen] = useState(false);
  const handleOpenerClick = (e) => {
    if (popoverRef.current) {
      popoverRef.current.opener = e.target;
      setOpen((prev) => !prev);
    }
  };
  return (
    <>
      <Button onClick={handleOpenerClick}>Opener</Button>
      <ActionSheet ref={popoverRef} open={open}>
        {/* Content */}
      </ActionSheet>
    </>
  );
};
```

#### before 0.22.0

`ActionSheets` can only be opened by attaching a `ref` to the component and then call the corresponding **`showAt`** method. The method receives the target element - _on which the `ActionSheet` is to be opened_ - as parameter.

```jsx
const ActionSheetComponent = () => {
  const actionSheetRef = useRef(null);
  const onButtonClick = (e) => {
    actionSheetRef.current.showAt(e.target);
  };
  return (
    <>
      <Button onClick={onButtonClick}>Open ActionSheet</Button>
      <ActionSheet ref={actionSheetRef}>
        <Button icon="add">Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    </>
  );
};
```

## Using ActionSheets inside other components

`ActionSheets` are often used within other components, when opened this could sometimes have unwanted side effects.
In this case, we recommend using [createPortal](https://reactjs.org/docs/portals.html) to mount the `ActionSheet` outside of the DOM hierarchy of the parent component.

```JSX
const ActionSheetComponent = () => {
  const actionSheetRef = useRef(null);
  const onButtonClick = (e) => {
    actionSheetRef.current.showAt(e.target);
  };
  return (
    <>
      <Button onClick={onButtonClick}>Open ActionSheet</Button>
      {createPortal(
        <ActionSheet ref={actionSheetRef}>
          <Button icon="add">Accept</Button>
          <Button>Reject</Button>
          <Button>This is my super long text!</Button>
        </ActionSheet>,
        document.body
      )}
    </>
  );
};
```

<Footer />
