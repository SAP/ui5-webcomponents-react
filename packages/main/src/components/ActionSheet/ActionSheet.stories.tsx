import { isChromatic } from '@sb/utils';
import type { Meta, StoryObj } from '@storybook/react';
import acceptIcon from '@ui5/webcomponents-icons/dist/accept.js';
import declineIcon from '@ui5/webcomponents-icons/dist/decline.js';
import deleteIcon from '@ui5/webcomponents-icons/dist/delete.js';
import emailIcon from '@ui5/webcomponents-icons/dist/email.js';
import forwardIcon from '@ui5/webcomponents-icons/dist/forward.js';
import React, { useState } from 'react';
import { PopoverHorizontalAlign, PopoverPlacementType, PopoverVerticalAlign } from '../../enums/index.js';
import { Button } from '../../webComponents/index.js';
import { ActionSheet } from './index.js';

const meta = {
  title: 'Modals & Popovers / ActionSheet',
  component: ActionSheet,
  argTypes: {
    children: {
      control: { disable: true }
    },
    footer: {
      control: { disable: true }
    },
    header: {
      control: { disable: true }
    },
    a11yConfig: { table: { category: 'A11y props' } }
  },
  args: {
    horizontalAlign: PopoverHorizontalAlign.Center,
    placementType: PopoverPlacementType.Right,
    verticalAlign: PopoverVerticalAlign.Center,
    open: false
  },
  parameters: {
    chromatic: { delay: 1000 }
  }
} satisfies Meta<typeof ActionSheet>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render(args) {
    const [actionSheetOpen, setActionSheetOpen] = useState(isChromatic || args.open);
    return (
      <div>
        <Button
          onClick={() => {
            setActionSheetOpen(true);
          }}
          id={'actionSheetOpener'}
        >
          Open ActionSheet
        </Button>
        <ActionSheet
          {...args}
          opener="actionSheetOpener"
          open={actionSheetOpen}
          onAfterClose={() => {
            setActionSheetOpen(false);
          }}
        >
          <Button icon={acceptIcon}>Accept</Button>
          <Button icon={declineIcon}>Reject</Button>
          <Button icon={emailIcon}>Email</Button>
          <Button icon={forwardIcon}>Forward</Button>
          <Button icon={deleteIcon}>Delete</Button>
          <Button>Other</Button>
        </ActionSheet>
      </div>
    );
  }
};
