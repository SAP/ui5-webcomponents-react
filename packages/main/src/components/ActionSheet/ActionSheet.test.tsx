import { fireEvent, render, screen } from '@shared/tests';
import { createPassThroughPropsTest } from '@shared/tests/utils';
import { ActionSheet } from '@ui5/webcomponents-react/dist/ActionSheet';
import { PopoverDomRef } from '@ui5/webcomponents-react/dist/Popover';
import { Button } from '@ui5/webcomponents-react/dist/Button';
import { Label } from '@ui5/webcomponents-react/dist/Label';
import React, { createRef, RefObject } from 'react';

describe('ActionSheet', () => {
  test('Render without Crashing', () => {
    const ref = createRef();
    const wrapper = render(
      <ActionSheet className="myCustomClass" ref={ref}>
        <Button>Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    );

    expect(wrapper.container.parentElement).toMatchSnapshot();
  });

  test('Button Click handler', () => {
    const callback = jest.fn();
    render(
      <ActionSheet className="myCustomClass">
        <Button onClick={callback}>Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    );

    fireEvent.click(screen.getByText('Accept'));

    expect(callback).toBeCalled();
  });

  test('Test Legacy Ref', () => {
    let legacyRef = null;

    const ref = (el) => {
      legacyRef = el;
    };
    render(
      <ActionSheet ref={ref}>
        <Button>Accept</Button>
        <Button>Reject</Button>
        <Button>This is my super long text!</Button>
      </ActionSheet>
    );
    expect(legacyRef.tagName).toEqual('UI5-RESPONSIVE-POPOVER');
  });

  test('Ref object', () => {
    const ref: RefObject<PopoverDomRef> = createRef();
    render(<ActionSheet ref={ref} />);
    expect((ref.current as any).tagName).toEqual('UI5-RESPONSIVE-POPOVER');
  });

  test('does not crash with other component', () => {
    const { container } = render(
      <ActionSheet>
        <Label>I should not crash</Label>
      </ActionSheet>
    );
    expect(container.parentElement).toMatchSnapshot();
  });

  createPassThroughPropsTest(ActionSheet);
});
