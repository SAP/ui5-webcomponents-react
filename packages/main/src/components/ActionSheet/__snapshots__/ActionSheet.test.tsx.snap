// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSheet Render without Crashing 1`] = `
<body>
  <div />
  <ui5-responsive-popover
    aria-modal="true"
    class="ActionSheet-actionSheet myCustomClass"
    data-actionsheet="true"
    horizontal-align="Center"
    placement-type="Right"
    role="dialog"
    ui5-responsive-popover=""
    vertical-align="Center"
  >
    <div
      aria-label="Available Actions"
      data-component-name="ActionSheetMobileContent"
      role="presentation"
    >
      <ui5-button
        aria-label="1 of 3 Accept"
        data-action-btn-index="0"
        design="Transparent"
        role="button"
        tabindex="0"
        ui5-button=""
      >
        Accept
      </ui5-button>
      <ui5-button
        aria-label="2 of 3 Reject"
        data-action-btn-index="1"
        design="Transparent"
        role="button"
        tabindex="-1"
        ui5-button=""
      >
        Reject
      </ui5-button>
      <ui5-button
        aria-label="3 of 3 This is my super long text!"
        data-action-btn-index="2"
        design="Transparent"
        role="button"
        tabindex="-1"
        ui5-button=""
      >
        This is my super long text!
      </ui5-button>
    </div>
  </ui5-responsive-popover>
</body>
`;

exports[`ActionSheet does not crash with other component 1`] = `
<body>
  <div />
  <ui5-responsive-popover
    aria-modal="true"
    class="ActionSheet-actionSheet"
    data-actionsheet="true"
    horizontal-align="Center"
    placement-type="Right"
    role="dialog"
    ui5-responsive-popover=""
    vertical-align="Center"
  >
    <div
      aria-label="Available Actions"
      data-component-name="ActionSheetMobileContent"
      role="presentation"
    >
      <ui5-label
        aria-label="1 of 1 I should not crash"
        data-action-btn-index="0"
        design="Transparent"
        role="button"
        tabindex="0"
        ui5-label=""
      >
        I should not crash
      </ui5-label>
    </div>
  </ui5-responsive-popover>
</body>
`;
